redis/src/cluster_async/mod.rs

```diff
@@ -288,11 +289,20 @@ where
                 request.retry = request.retry.saturating_add(1);
 
                 match err.kind() {
-                    // TODO: Proper Ask implementation
-                    ErrorKind::Moved | ErrorKind::Ask => Next::RefreshSlots {
-                        request: this.request.take().unwrap(),
+                    ErrorKind::Ask => {
+                        let mut request = this.request.take().unwrap();
+                        request.info.redirect = err
+                            .redirect_node()
+                            .map(|(node, _slot)| Redirect::Ask(node.to_string()));
+                        Next::Retry { request }.into()
+                    }
+                    ErrorKind::Moved => {
+                        let mut request = this.request.take().unwrap();
+                        request.info.redirect = err
+                            .redirect_node()
+                            .map(|(node, _slot)| Redirect::Moved(node.to_string()));
+                        Next::RefreshSlots { request }.into()
```

Comment:
not to fix in this PR: if I understand the code correctly, this result is queued into `in_flight_requests`. This means that if we decide to refresh slots, but there are other requests on there, the slot refresh will be delayed. Should we add some kind of mechanism to give slot refresh preference, and run that first?

Reply:
Agreed. I don't think it's a huge concern at the moment but would be ideal if we can do so w/out further complicating the code.